### ISSUE TITLE
Channel membership and channel context lifecycle

### ISSUE THREAD

**Issue by nicholasdgoodman** [finos/FDC3#162]
## Feature Request
**TL;DR**
- An app may join multiple channels at once
- A `DesktopAgent` may select exactly one Channel as its context source

(Sorry this is seemingly contrary to my comments on PR https://github.com/finos/FDC3/pull/158)

### Description of Problem:

The current `Channel` API proposal allows for holding onto references of unjoined channels and creating custom app channels. Take for example the following code:

```
let redChannel = (await fdc3.getSystemChannels()).find(ch => ch.id === 'red');
let greenChannel = (await fdc3.getSystemChannels()).find(ch => ch.id === 'green');
let appChannel = await fdc3.getOrCreateChannel('appChannel');

fdc3.join(redChannel.id);
```

Based on the existing specification, having references to these channels is sufficient for most of their functionality: `broadcast(...)`, `getCurrentContext(...)`, and presumably `addContextListener(...)` all function regardless of whether or not the channels have been joined.

However, in the current spec, context persistence only occurs when there is one or more application which has joined the channel. By design, channels lose their state / context when the last member leaves (or when no members are present).

Consider:

```
// this channel has no members
appChannel.broadcast({type: 'fdc3.contact'});

// what should the context be here?
let appChannelContext = await appChannel.getCurrentContext();
```

Presumably, the context of a channel with no members remains `undefined`; although the API provides no means of discerning this.

Furthermore, it seems unnecessarily restrictive to limit an app to joining a single channel when it can hold references, query status, and receive events from multiple channels.

### Potential Solutions:

Allow an app to explicitly join or leave a channel from the `Channel` object itself. This membership has no affect on the context source of the `DesktopAgent`:

```
// Only affects the membership list of the channel
// allows it to persist state:
Channel.join() : Promise<void>

// If the app is the last member of the channel,
// leaving it causes the context to be cleared:
Channel.leave() : Promise<void>

// Denotes if a channel has members
// and will thus persist context
Channel.hasMembers() : Promise<bool>

// Designates a specified channel provides the event source
// for the DesktopAgent. Joins the channel if it is not joined
// By default, leaves the currently joined channel
// Channel.contextListeners bubble to DesktopAgent.contextListeners
DesktopAgent.joinChannel(channelId: string, leaveCurrent: boolean = true);
```

**Comment by nkolba** [comment]
@nicholasdgoodman - Not sure if I totally understand the use case here,  but it seems like the main issue your solving for is that : *in the current spec, context persistence only occurs when there is one or more application which has joined the channel* . 

I don't believe this is explicitly in the spec.  There may be a behavior in implementations to drop the context if all apps have left a channel.  The use case for this with system channels would be to mitigate the potential for unexpected behavior.  If, for example, I'm using the red channel, close everything that was on the red channel, then open a new set of apps and put them on the red channel.  In this case, I may not expect to get my old context.   I agree that this needs to be written up more carefully in the spec.

A main use case for the single channel membership rule is to allow channel membership for apps to be managed by the end user (and outside of the control of the app).  This means that an FDC3 implementor can just put in logic for:

```js
fdc3.addContextListener(...)
fdc3.broadcast(...)
```

and not be aware of channels at all for use cases of color linking or similar use cases of apps being implicitly linked in a workspace.  Everything else is possible via apps using the `addContextListener/broadcast` methods on the Channel objects themselves.

**Comment by nicholasdgoodman** [comment]
@nkolba thats good news. after re-reading the proposed spec, i can see now that there is no behavior specified in the case of channels that have no Agents / apps joined to them. this was a misunderstanding on my part as i had previously been looking only at one particular (pre-spec) implementation of channels (i think you know where 😁) and that was the behavior i observed.

if channel objects are fully functional as standalone objects, then i have no other concerns.

from an implementors perspective, it does feel like there should be some form of lifecycle management around the channels (they can be created but not destroyed), though such a concern need not be covered in the spec, IMO
